### PR TITLE
Eliminate `UTILS_UNUSED_PARAM(..)` macro

### PR DESCRIPTION
**Pull request by ngie-eign**
Unlike C, C++ has the ability to omit parameter names in implemented
interfaces when the parameter is unused, aka "unnamed parameters"

This change does the following:
1. Mechanically replace `UTILS_UNUSED_PARAM(x)` with `/* x */`.
2. Removes all doxygen directives that reference `unused_*`.
3. Removes the m4 macro that looks for `__attribute__((unused))`.

Signed-off-by: Enji Cooper <yaneurabeya@gmail.com>

### CODE REVIEW

**Comment by ngie-eign** [comment]
@jmmv: this came about as a subthread when reviewing #197. It probably won't pass Travis because Travis is broken on master. #195 would make this change a bit more green by at least checking to make sure things compile on Travis.

File: bootstrap/plain_helpers.cpp [hunk 1]
```diff
@@ -59,26 +58,22 @@ helper_abort_test_cases_list(int UTILS_UNUSED_PARAM(argc), char** argv)
...
 ///
-/// \param unused_argc The original argument count of the program.
-/// \param unused_argv The original arguments of the program.
-///
 /// \return Always 0, as required for test programs.
 static int
-helper_empty_test_cases_list(int UTILS_UNUSED_PARAM(argc),
-                             char** UTILS_UNUSED_PARAM(argv))
+helper_empty_test_cases_list(int /* argc */,
+                             char** /* argv */)
```
Comment (jmmv) [inline bootstrap/plain_helpers.cpp]: These now fit in a single line?

File: drivers/report_junit.cpp [hunk 2]
```diff
@@ -251,11 +251,9 @@ drivers::report_junit_hooks::got_result(store::results_iterator& iter)
 
 
 /// Finalizes the report.
-///
-/// \param unused_r The result of the driver execution.
 void
 drivers::report_junit_hooks::end(
-    const drivers::scan_results::result& UTILS_UNUSED_PARAM(r))
+    const drivers::scan_results::result& /* r */)
```
Comment (jmmv) [inline drivers/report_junit.cpp]: One line maybe?

File: model/metadata.cpp [hunk 3]
```diff
@@ -81,21 +81,16 @@ class bytes_node : public config::native_leaf_node< units::bytes > {
...
 
     /// Sets the value of the node from an entry in the Lua stack.
-    ///
-    /// \param unused_state The Lua state from which to get the value.
-    /// \param unused_index The stack index in which the value resides.
     void
-    set_lua(lutok::state& UTILS_UNUSED_PARAM(state),
-            const int UTILS_UNUSED_PARAM(index))
+    set_lua(lutok::state& /* state */,
+            const int /* index */)
```
Comment (jmmv) [inline model/metadata.cpp]: One line.

File: model/metadata.cpp [hunk 4]
```diff
@@ -145,21 +140,16 @@ class delta_node : public config::typed_leaf_node< datetime::delta > {
...
     }
 
     /// Sets the value of the node from an entry in the Lua stack.
-    ///
-    /// \param unused_state The Lua state from which to get the value.
-    /// \param unused_index The stack index in which the value resides.
     void
-    set_lua(lutok::state& UTILS_UNUSED_PARAM(state),
-            const int UTILS_UNUSED_PARAM(index))
+    set_lua(lutok::state& /* state */,
```
Comment (jmmv) [inline model/metadata.cpp]: One line.

File: utils/config/nodes.ipp [hunk 5]
```diff
@@ -217,13 +217,11 @@ config::typed_leaf_node< ValueType >::set(const value_type& value_)
...
 /// value.  This method can be redefined by subclasses.
 ///
-/// \param unused_new_value The value to validate.
-///
 /// \throw value_error If the value is not valid.
 template< typename ValueType >
 void
 config::typed_leaf_node< ValueType >::validate(
-    const value_type& UTILS_UNUSED_PARAM(new_value)) const
+    const value_type& /* new_value */) const
```
Comment (jmmv) [inline utils/config/nodes.ipp:224]: One line maybe?

File: utils/config/nodes.ipp [hunk 6]
```diff
@@ -367,29 +365,24 @@ config::base_set_node< ValueType >::to_string(void) const
 
 
 /// Pushes the node's value onto the Lua stack.
-///
-/// \param unused_state The Lua state onto which to push the value.
 template< typename ValueType >
 void
 config::base_set_node< ValueType >::push_lua(
-    lutok::state& UTILS_UNUSED_PARAM(state)) const
+    lutok::state& /* state */) const
```
Comment (jmmv) [inline utils/config/nodes.ipp]: One line maybe?

File: utils/optional.ipp [hunk 7]
```diff
@@ -45,11 +45,9 @@ utils::optional< T >::optional(void) :
 
 
 /// Explicitly initializes an optional object to the none value.
-///
-/// \param unused_none A copy of the utils::none instance.
 template< class T >
 utils::optional< T >::optional(
-    utils::detail::none_t UTILS_UNUSED_PARAM(none)) :
+    utils::detail::none_t /* none */) :
     _data(NULL)
```
Comment (jmmv) [inline utils/optional.ipp:50]: One line for everything?

File: utils/process/executor_test.cpp [hunk 8]
```diff
@@ -162,12 +159,9 @@ static void child_dump_unprivileged_user(const fs::path&) UTILS_NORETURN;
 
 
 /// Subprocess that dumps user configuration.
-///
-/// \param unused_control_directory Directory where control files separate from
-///     the work directory can be placed.
 static void
 child_dump_unprivileged_user(
-    const fs::path& UTILS_UNUSED_PARAM(control_directory))
+    const fs::path& /* control_directory */)
```
Comment (jmmv) [inline utils/process/executor_test.cpp]: One line.

File: utils/sqlite/statement.cpp [hunk 9]
```diff
@@ -543,12 +543,11 @@ sqlite::statement::bind(const int index, const int64_t value)
 /// Binds a NULL value to a prepared statement.
 ///
 /// \param index The index of the binding.
-/// \param unused_null An instance of the null class.
 ///
 /// \throw api_error If the binding fails.
 void
 sqlite::statement::bind(const int index,
-                             const null& UTILS_UNUSED_PARAM(null))
+                             const null& /* null */)
```
Comment (jmmv) [inline utils/sqlite/statement.cpp]: Bad indentation, but fix in one line?

File: utils/defs.hpp.in [hunk 10]
```diff
@@ -43,28 +43,11 @@
 #define UTILS_PURE @ATTRIBUTE_PURE@
 
 
-/// Attribute to mark an entity as unused.
-#define UTILS_UNUSED @ATTRIBUTE_UNUSED@
```
Comment (jmmv) [inline utils/defs.hpp.in]: I'm surprised to see this go, given that this is different than the parameter issue and it's still used. Why?

**Review by jmmv — COMMENTED** [review]
Everything just trivial formatting issues except for one real comment.

File: utils/config/nodes.ipp (hunk 5 above)
Comment (ngie-eign) [reply utils/config/nodes.ipp:224]: Unfortunately this doesn't fit :/.

File: utils/defs.hpp.in (hunk 10 above)
Comment (ngie-eign) [reply utils/defs.hpp.in]: Ah, shoot.. I didn't realize that :/.

**Comment by ngie-eign** [comment]
@jmmv: ready! It needs some more fix up in order to pass `make apidocs`, but I will resolve that as part of #197.

**Comment by ngie-eign** [comment]
@jmmv: ping?

**Comment by ngie-eign** [comment]
This was merged as part of #197, so I'm closing this PR.

**Comment by jmmv** [comment]
Was it? The #197 diff doesn't mention UTILS_UNUSED_PARAM at all...

**Comment by ngie-eign** [comment]
> Was it? The #197 diff doesn't mention UTILS_UNUSED_PARAM at all...

It was based on that branch. Sorry about the confusion >_>..

**Comment by jmmv** [comment]
I don't understand... even if that commit was on top of another unmerged branch, GitHub should have showed the diff against the master branch (and thus have showed the two changes). Also, looking at head, I still see the macro. So... this PR needs to be revived? Is that what you meant?

**Comment by ngie-eign** [comment]
> I don't understand... even if that commit was on top of another unmerged branch, GitHub should have showed the diff against the master branch (and thus have showed the two changes). Also, looking at head, I still see the macro. So... this PR needs to be revived? Is that what you meant?

Oh wait! I split them up! Shoot... yeah, I'll revive the branch in a sec.

**Comment by ngie-eign** [comment]
Ugh. Need to use another name. GitHub isn't letting me reopen this because I prematurely deleted the branch :(.

**Comment by ngie-eign** [comment]
Ok, opened a new PR: #201. I'm waiting for it to complete.

**Pull request closed without merging**